sklego/pandas.py

```diff
@@ -0,0 +1,186 @@
+import numpy as np
+import pandas as pd
+from scipy.ndimage.interpolation import shift
+
+
+def _as_list(val):
+    """
+    Helper function, always returns a list.
+
+    :param val: a list or a single value.
+    :returns: the input value as a list.
+
+    :Example:
+
+    >>> _as_list('test')
+    ['test']
+
+    >>> _as_list(['test1', 'test2'])
+    ['test1', 'test2']
+    """
+
+    if not isinstance(val, list):
+        return [val]
+    return val
+
+
+def _negate_lags(lags):
+    """
+    Helper function, negates an int or iterable of ints.
+
+    :param lags: iterable of ints or int.
+    :returns: iterable of ints that are negated.
+
+    :Example:
+
+    >>> _negate_lags(3)
+    [-3]
+
+    >>> _negate_lags([3])
+    [-3]
+    """
+
+    def negate(x):
+        return -x
+
+    lags = _as_list(lags)
+
+    if not all(isinstance(x, int) for x in lags):
+        raise ValueError("Must be list of ints")
+
+    return [-x for x in lags]
+
+
+def add_lag(X, cols, lags):
+    """
+    Appends lag column(s).
+
+    :param X: array-like, shape=(n_columns, n_samples,) training data.
+    :param cols: an iterable of column names or a single column name string.
+    :param lags: iterable of ints or int.
+    :returns: ``pd.DataFrame`` with only the selected columns
+
+    :Example:
+
+    >>> import pandas as pd
+    >>> df = pd.DataFrame([[1, 2, 3],
+    ...                    [4, 5, 6],
+    ...                    [7, 8, 9]],
+    ...                    columns=['a', 'b', 'c'],
+    ...                    index=[1, 2, 3])
+
+    >>> add_lag(df, 'a', [1]) # doctest: +NORMALIZE_WHITESPACE
+    a  b  c  a-1
+    1  1  2  3  4.0
+    2  4  5  6  7.0
+
+    >>> add_lag(df, ['a', 'b'], 2) # doctest: +NORMALIZE_WHITESPACE
+       a  b  c  a-2  b-2
+    1  1  2  3  7.0  8.0
+
+    >>> import numpy as np
+    >>> X = np.array([[-4, 2],
+    ...               [-2, 0],
+    ...               [4, -6]])
+
+    >>> add_lag(X, [0, 1], 1)
+    array([[-4,  2, -2,  0],
+           [-2,  0,  4, -6],
+           [ 4, -6,  0,  0]])
+
+    >>> add_lag(X, 1, [-1, 1])
+    array([[-4,  2,  0,  0],
+           [-2,  0,  2, -6],
+           [ 4, -6,  0,  0]])
+    """
+
+    lags = _negate_lags(lags)
+
+    # Check the X instance of pandas DataFrame
+    if isinstance(X, pd.core.frame.DataFrame):
+        return _add_lagged_pandas_columns(X, cols, lags)
+
+    # Check the X instance of numpy array
+    if isinstance(X, np.ndarray):
+        return _add_lagged_numpy_columns(X, cols, lags)
+
+    raise ValueError("Please pass a pandas DataFrame")
+
+
+def _add_lagged_numpy_columns(X, cols, lags):
```

Comment:
if this is a hidden function to the user, why write an example for it?

Reply:
In my opinion, if you decide in the future to make it public, it's trivial. If a complex public function becomes private, would you remove the examples?

It also documents the intention for those who may have to debug it, and it works as a proof that everything is functioning correctly, as otherwise you'd have to rely on testing multiple cases in the function that calls the hidden function. For example:
- does numpy work with input x?
- does numpy work with input y?
- does numpy fail correctly with a wrong input?
- does pandas work with input x?
- does pandas work with input y?
- does pandas fail correctly with a wrong input?

Would you keep adding tests to 'add_lag' just to make sure the underlying hidden functions are working as intended?

Reply:
Nothing wrong with writing tests for it but in this case I would indeed put them in a separate unit test rather than in doctests

Reply:
Okay, so only the description and params sections of the docstring of hidden functions should remain?